### DISCUSSION TITLE
How to output OpenAPI spec with custom key need for advanced code generation features like security

### DISCUSSION THREAD

**Discussion by luistrigueiros** [microsoft/typespec#3364] [OpenAPI3] [answered]
In my use case I would like to output OpenAPI spec but I am looking I would need for the output YAML file to contain extra key's that are hint's to the code generator that allow for advanced features in the generated code like support for SpringSecurity annotations in the web controller.
How can this be achieved with TypeSpec ?
The example output desired can be found here: https://github.com/OpenAPITools/openapi-generator/pull/16259/files#diff-c70ab8ecddee5285186949a1ec4710e8dc05c7176cf4c47c209abb92a02ec809R31

`
      x-operation-extra-annotation:
        - '@javax.annotation.security.RolesAllowed({"ROLE_TEST"})'
        - '@org.springframework.security.access.annotation.Secured({"ROLE_TEST"})'
        - '@org.springframework.security.access.prepost.PreAuthorize("hasRole(''ROLE_TEST'') and hasRole(''ROLE_TEST_TWO'')")'
`

![screenshot 2](https://github.com/microsoft/typespec/assets/590567/f75273cc-3949-4b25-a8ac-312c986ca8e3)

Thank you, kindly
   Luis Oscar  Trigueiros

**Comment by timotheeguerin** [comment, marked as answer]
You can use the `@OpenAPI.extension` decorator from the `@typespec/openapi` library 

https://typespec.io/playground?c=aW1wb3J0ICJAdHlwZXNwZWMvb3BlbmFwaSI7Cgp1c2luZyBPcGVuQVBJOwoKQGV4dGVuc2lvbigKICAieC1vcGVyYXRpb24tZXh0cmEtYW5ub3TFESIsCiAgWwogICAgIkBqYXZheC7KHS5zZWN1cml0eS5Sb2xlc0FsbG93ZWQoe1wiUk9MRV9URVNUXCJ9KcVExEBvcmcuc3ByaW5nZnJhbWV3b3JrykNhY2Nlc3PMXlPEG99Q31BzLnByZXBvc3QuUHJlQXV0aG9yaXplKFwiaGFz5AC0KCfJWicpIGFuZCDSGV9UV08nKVwiKSIKICBdCikKb3AgdGVzdCgpOiB2b2lkOwo%3D&e=%40typespec%2Fopenapi3&options=%7B%7D

**Reply by luistrigueiros** [reply]
Thank you, for the answer, wow I am impressed by the link to the playground with the generated spec

**Comment by bpottier** [comment]
Is there a way to do this globally? We have a use case where we want to apply an extension to all operations and want to avoid writing the same decorator all over the place.

**Reply by timotheeguerin** [reply]
You cannot do that today, the recommended approach is to create reusable operation templates this way [playground](https://typespec.io/playground?c=aW1wb3J0ICJAdHlwZXNwZWMvaHR0cCI7CtIZb3BlbmFwaSI7Cgp1c2luZyBPcGVuQVBJO8cPSHR0cDsKCkBleHRlbnNpb24oCiAgIngtb3BlcmF0aW9uLWV4dHJhLWFubm90xREiLAogIFsKICAgICJAamF2YXguyh0uc2VjdXJpdHkuUm9sZXNBbGxvd2VkKHtcIlJPTEVfVEVTVFwifSnFRMRAb3JnLnNwcmluZ2ZyYW1ld29ya8pDYWNjZXNzzF5TxBvfUN9Qcy5wcmVwb3N0LlByZUF1dGhvcml6ZShcImhhc%2BQAtCgnyVonKSBhbmQg0hlfVFdPJylcIikiCiAgXQopCm9wIEFj5ACiPFBhcmFtcywgT3V0cHV0PiguLi7GEyk6xxQ7CsQvbGlzdCBpc8g3e30sIHZvaWQ%2BO8QdZ2XMHAogIHvlAN9AcGF0aCBpZDogc3TkAOM7CiAgfeQA%2BsQ6Cj47Cg%3D%3D&e=%40typespec%2Fopenapi3&options=%7B%7D)

**Comment by KnisterPeter** [comment]
@timotheeguerin I found this and at least I could work with it. I want to emit extension information for an AWS ApiGateway. But it's quite verbose and repetititve to do it with decorators like this one.
Can I create a 'meta-decorator' to reduce the boilerplate?

**Reply by KnisterPeter** [reply]
Would it be safe to import the `setExtension` method from the openapi package and call it from my own decorator?
